### PR TITLE
Enhance/dropdown align right

### LINKED ISSUE #1257: [Enhancement] Allow dropdown to optionally align end
**Is your enhancement request related to a problem? Please describe.**
There are specific situations where dropdown needs to align end, for example when used in calendar. In calendar the dropdown caret is placed at the end of the content, and so when you use the dropdown it overlaps some of the content but mostly half of the content.

image

**Describe the solution you'd like**
In this situation we would like the dropdown to align with the calendar by setting a property exposing the alignment of the dropdown through the API.

**Describe alternatives you've considered**
An alternative is assigning a utility css class, however since we prefer developers to consume kirby through its API, hence making them explicitly aware of the ability to choose alignment, the suggested solution is the way to go.

**Additional context**
This is the desired result:
image

### PR DESCRIPTION
**Pull request by flemcito**
This PR closes #1257

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

This PR introduces the option to declaratively force the dropdown card to be right-aligned, related to the button

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

The dropdown card is currently always left-aligned unless the card is cutoff, ie. if the dropdown is right-aligned and the card is too wide to fit inside the surrounding container. Then the dropdown will automatically change the card alignment to right.

## What is the new behavior?

It is now possible to reverse the current behaviour by adding a `open="left"` attribute. This will make the expanding card right-aligned and reverse the current overflow check - so the card will be left-aligned if cut off.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 1]
```diff
@@ -253,6 +258,33 @@ describe('DropdownComponent', () => {
       });
     });
 
+    fdescribe('when configured with cardAligment', () => {
```
Comment (jkaltoft) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:261]: `fdescribe` --> `describe` without the `f` prefix

File: .tool-versions [hunk 2]
```diff
@@ -0,0 +1 @@
+nodejs 12.20.0
```
Comment (jakobe) [inline .tool-versions:1]: Is this intentionally added? If so, seems out of scope for this PR - if not, please revert :)

File: apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { Component, Input } from '@angular/core';
+
+const config = {
+  selector: 'cookbook-dropdown-example-dropdown-right-aligned',
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts:4]: Consider naming:
```suggestion
  selector: 'cookbook-dropdown-example-direction',
```
OR
```suggestion
  selector: 'cookbook-dropdown-example-open-left',
```

File: apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html [hunk 4]
```diff
@@ -8,6 +8,9 @@ <h3>Simple</h3>
 <cookbook-dropdown-example-pre-selected [size]="size"></cookbook-dropdown-example-pre-selected>
 <cookbook-dropdown-example-expand [size]="size"></cookbook-dropdown-example-expand>
 <cookbook-dropdown-example-right-aligned [size]="size"></cookbook-dropdown-example-right-aligned>
+<cookbook-dropdown-example-dropdown-right-aligned
+  [size]="size"
+></cookbook-dropdown-example-dropdown-right-aligned>
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html:13]: Consider if this example could be merged with existing right-aligned example (showcasing right alignment + open to the left in one example)

File: apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  template: `<kirby-dropdown
+  [size]="size"
+  placeholder="Right aligned dropdown"
+  [items]="['Item 1','Item 2','Item 3','Item 4','Item 5']"
+  cardAlign="right"
+></kirby-dropdown>`,
+  styles: [
+    `kirby-dropdown {
+margin-left: 200px;
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts:13]: Will this work on small screens? 🤔

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 6]
```diff
@@ -253,6 +258,33 @@ describe('DropdownComponent', () => {
...
 
+    describe('when configured with cardAligment', () => {
+      it('should render card with left alignment', () => {
+        spectator.component.cardAlign = cardAlignment.left;
+        spectator.component['state'] = openState.open;
+        spectator.detectChanges();
+        const buttonRect = buttonElement.getBoundingClientRect();
+        const card = spectator.query('kirby-card');
+        const cardRect = card.getBoundingClientRect();
+        expect(cardRect.left).toEqual(buttonRect.left);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:269]: Consider adding linebreaks in a arrange-act-assert pattern:
```suggestion
        spectator.component.cardAlign = cardAlignment.left;
        spectator.component['state'] = openState.open;
        spectator.detectChanges();
        
        const buttonRect = buttonElement.getBoundingClientRect();
        const card = spectator.query('kirby-card');
        const cardRect = card.getBoundingClientRect();
        expect(cardRect.left).toEqual(buttonRect.left);
```

Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:269]: Nice! 👌

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 7]
```diff
@@ -253,6 +258,33 @@ describe('DropdownComponent', () => {
...
+        spectator.component.open();
+        spectator.detectChanges();
+        setTimeout(() => {
+          spectator.detectChanges();
+          const card = spectator.query('kirby-card');
+          const buttonRect = buttonElement.getBoundingClientRect();
+          const cardRect = card.getBoundingClientRect();
+          expect(cardRect.right).toEqual(buttonRect.right);
+          done();
+        }, openDelayInMs);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:284]: Consider adding linebreaks in a arrange-act-assert pattern:
```suggestion
        spectator.setInput('cardAlign', cardAlignment.right);
        spectator.element.style.cssFloat = 'right';
        spectator.component.open();
        spectator.detectChanges();
        
        setTimeout(() => {
          spectator.detectChanges();
          const card = spectator.query('kirby-card');
          const buttonRect = buttonElement.getBoundingClientRect();
          const cardRect = card.getBoundingClientRect();
          expect(cardRect.right).toEqual(buttonRect.right);
          done();
        }, openDelayInMs);
```

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 8]
```diff
@@ -253,6 +258,33 @@ describe('DropdownComponent', () => {
...
+        expect(cardRect.left).toEqual(buttonRect.left);
+      });
+
+      it('should render card with right alignment', (done) => {
+        spectator.setInput('cardAlign', cardAlignment.right);
+        spectator.element.style.cssFloat = 'right';
+        spectator.component.open();
+        spectator.detectChanges();
+        setTimeout(() => {
+          spectator.detectChanges();
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:278]: Why the duplicated change detection? 🤔

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 9]
```diff
@@ -253,6 +258,33 @@ describe('DropdownComponent', () => {
...
+        const cardRect = card.getBoundingClientRect();
+        expect(cardRect.left).toEqual(buttonRect.left);
+      });
+
+      it('should render card with right alignment', (done) => {
+        spectator.setInput('cardAlign', cardAlignment.right);
+        spectator.element.style.cssFloat = 'right';
+        spectator.component.open();
+        spectator.detectChanges();
+        setTimeout(() => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:277]: Why not just setting the state to open as in the other test? 🤔

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 10]
```diff
@@ -1,30 +1,40 @@
...
-import { ItemComponent } from '../item/item.component';
 import { CardComponent } from '../card/card.component';
+import { ItemComponent } from '../item/item.component';
+import { ListItemTemplateDirective } from '../list/list.directive';
 
+export enum openState {
+  closed,
+  opening,
+  open,
+}
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:32]: Could you comment on why this is better than the string union type? 🤔

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 11]
```diff
@@ -71,6 +81,9 @@ export class DropdownComponent
   @Input()
   placeholder = 'Please select:';
 
+  @Input()
+  cardAlign: cardAlignment = cardAlignment.left;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:85]: Consider naming - would it be better/easier to understand if this was named according to the direction the card/dropdown opens in? E.g. `direction` or `openDirection` or simply `open`, i.e.: `open="left"`
Ionic uses `vertical` and `horizontal` for their FABs: https://ionicframework.com/docs/api/fab#properties
(that API also takes into account to set both the vertical and horizontal direction, whereas only having `direction` would either only support one "compass" direction (NESW) or need to to contain  a matrix of horizontal/vertical aligments if ever needed.... 🤔)
UPDATE: I see now that we already have  `_vertical` and `_horizontal` properties 😂

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 12]
```diff
@@ -217,22 +230,20 @@ export class DropdownComponent
...
         const entry = entries[0];
         const isVisible = entry.boundingClientRect.width > 0;
         if (isVisible && entry.intersectionRatio < 1) {
-          // entry not fully showing:
-          if (entry.boundingClientRect.right > entry.rootBounds.right) {
-            // entry is cut off to the right by ${entry.boundingClientRect.right - entry.intersectionRect.right}px
-            // align to the end:
-            this._horizontal = 'end';
-          }
+          this.setCardAlignment(entry);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:245]: Nice refactoring 👌 
Consider if vertical aligment should be extracted to a method as well....

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 13]
```diff
@@ -257,12 +268,28 @@ export class DropdownComponent
     }
   }
 
+  private setCardAlignment(entry) {
+    // If card alignment is left, and the entry is cut off to the right by ${entry.boundingClientRect.right - entry.intersectionRect.right}px
+    // it is set to align to end in stead, and vice versa for right-aligned card
+    if (this.cardAlign === cardAlignment.left) {
+      if (entry.boundingClientRect.right > entry.rootBounds.right) {
+        this._horizontal = 'end';
+        this.cardAlign = cardAlignment.right;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:277]: Maybe these to lines are hinting at wanting to be the same thing/concept as mentioned in the comment regarding `cardAlign` naming? 🤔

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 14]
```diff
@@ -257,12 +268,28 @@ export class DropdownComponent
     }
   }
 
+  private setCardAlignment(entry) {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:271]: Consider naming:
```suggestion
  private setHorizontalAlignment(entry) {
```

File: libs/designsystem/testing-base/src/lib/components/mock.dropdown.component.ts [hunk 15]
```diff
@@ -1,6 +1,8 @@
-import { forwardRef, Component, Input, Output, EventEmitter } from '@angular/core';
-
-import { DropdownComponent } from '@kirbydesign/designsystem';
+import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
+import {
+  cardAlignment,
+  DropdownComponent,
+} from 'libs/designsystem/src/lib/components/dropdown/dropdown.component';
```
Comment (jakobe) [inline libs/designsystem/testing-base/src/lib/components/mock.dropdown.component.ts:5]: Please import from lib namespace (add `cardAlignment` to `components/index.ts` barrel file if needed) :
```suggestion
} from '@kirbydesign/designsystem';

```

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 16]
```diff
@@ -1,30 +1,40 @@
...
 } from '@angular/core';
 import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
 
-import { ListItemTemplateDirective } from '../list/list.directive';
-import { ItemComponent } from '../item/item.component';
 import { CardComponent } from '../card/card.component';
+import { ItemComponent } from '../item/item.component';
+import { ListItemTemplateDirective } from '../list/list.directive';
 
+export enum openState {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:28]: Please align enum naming convention with other enums in lib (whether that's the right casing might be a discussion it itself, but for now consistency is king 👑 😄 ) :
```suggestion
enum OpenState {
```

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 17]
```diff
@@ -1,30 +1,40 @@
...
+  closed,
+  opening,
+  open,
+}
+
+export enum cardAlignment {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:34]: Please align enum naming convention with other enums in lib (whether that's the right casing might be a discussion it itself, but for now consistency is king 👑 😄 ) :
```suggestion
export enum CardAlignment {
```

**Review by jakobe — COMMENTED** [review]
Nice work 👌 🎉 
My comments are mostly related to the API/naming of the property used to align the card/direction the dropdown opens in - let's have a talk around that 👍

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 1 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Doh.. removed :)

File: apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts (hunk 3 above)
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts:4]: I like the `open-left` - it's most descriptive 👍

File: apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html (hunk 4 above)
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html:13]: I'll do that. The reason I opted not to, was that the right aligned example triggers the interception observer and is as such demo-ing another implementation detail. But I guess that is kind of irrelevant for the cookbook consumer

File: apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts (hunk 5 above)
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts:13]: Good point

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 10 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:32]: Sure. There are a number of things :)
1. I think it is much cleaner and easier to reason about to have explicit types like this in stead of just strings
2. Reusability, the string-union types need to be declared every time whereas the enums can live in a central place an be consistent across the solution (ie. attentionLevel and sizes are excellent candidates for that)
3. Better tooling support - type a `.` and get all options
4. String literals are objectively bad :)
5. Union types is IMO an anti-pattern (though some times necessary), as it inherently undermines the main reason you would adopt typescript over vanilla JS.
6. The string-union type IS essentially an emulation of an enum rather than a union-type - IMO better to use the proper language construct

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 11 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:85]: Ha.. I actually though that `_horizontal` was an inherited prop. Totally missed that.
I will re-arrange a bit to not re-invent the wheel too many times

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 8 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:278]: Copied the timeout from another test that had it duplicated  🤦

File: apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html (hunk 4 above)
Comment (jakobe) [reply apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html:13]: Valid point 👍 
But on larger screens (ie. room for default opening direction) we get this 👇  which isn't demoing the right alignment every well anyhoo 🙈 
image

File: .tool-versions (hunk 2 above)
Comment (flemcito) [reply .tool-versions]: removed. I think we should consider adding this, but you are right - it's out of scope here

File: apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html (hunk 4 above)
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/dropdown-example.component.html]: Done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 6 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 14 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Renamed

File: libs/designsystem/testing-base/src/lib/components/mock.dropdown.component.ts (hunk 15 above)
Comment (flemcito) [reply libs/designsystem/testing-base/src/lib/components/mock.dropdown.component.ts]: Done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 16 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 17 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Renamed following convention

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 13 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Refactored

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 8 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Removed duplicate

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 9 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Also a copy/paste.. made it consistent

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 11 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: refactored

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 12 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Also extracted vertical

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 7 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Done

File: apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts (hunk 5 above)
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/examples/dropdown-right-aligned.ts]: Removed the separate cookbook entry -solving this

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 18]
```diff
@@ -268,28 +264,44 @@ export class DropdownComponent
...
+      // Check if the card can fit on top of button:
+      if (containerOffsetTop > entry.target.clientHeight + SPACING) {
+        // open upwards:
+        this._vertical = VerticalDirection.up;
       }
     }
   }
 
-  open() {
+  openDropdown() {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:299]: Unfortunately `open()` is part of the public and published API, so shouldn't change (or at least have a very good reason to...)
 I think we'll have to keep it and rename the direction input to `openDirection` or similar...

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 19]
```diff
@@ -299,25 +311,25 @@ export class DropdownComponent
...
-      this.state = openState.open;
+    if (this.state === OpenState.opening) {
+      this.state = OpenState.open;
       this.scrollItemIntoView(this.selectedIndex);
       this.changeDetectorRef.markForCheck();
     }
   }
 
-  close() {
+  closeDropdown() {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:321]: Unfortunately `close()` is part of the public and published API, so shouldn't change (or at least have a very good reason to...)

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 20]
```diff
@@ -268,28 +264,44 @@ export class DropdownComponent
...
     } else {
       if (entry.boundingClientRect.left < entry.rootBounds.left) {
-        this._horizontal = 'start';
-        this.cardAlign = cardAlignment.left;
+        this._horizontal = HorizontalDirection.right;
+      }
+    }
+  }
+
+  private setVerticalDirection(entry) {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:281]: Nice 👌

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 21]
```diff
@@ -144,25 +154,29 @@ export class DropdownComponent
...
+    this._openLeft = alignment === HorizontalDirection.left;
+  }
+  private get _horizontal() {
+    return this._openLeft ? HorizontalDirection.left : HorizontalDirection.right;
   }
-  private set _vertical(value: 'up' | 'down') {
-    this._alignTop = value === 'up';
+
+  private set _vertical(direction: VerticalDirection) {
+    this._openUp = direction === VerticalDirection.up;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Consider flipping this and using `_horizontal` and `_vertical` as single-source-of-truth.
The HostBinding property can then be used as a boolean as to whether the css class should be applied or not:
```ts
private _horizontal: HorizontalDirection;

@HostBinding('class.align-end')
  get _openLeft() {
    return _horizontal === HorizontalDirection.left;
  }
```

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 22]
```diff
@@ -144,25 +154,29 @@ export class DropdownComponent
...
 
   @HostBinding('class.is-open')
   get isOpen(): boolean {
-    return this.state === openState.open;
+    return this.state === OpenState.open;
   }
 
   @HostBinding('class.align-end')
-  _alignEnd: boolean;
+  _openLeft: boolean;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:166]: Consider align class-name if renaming property...

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 23]
```diff
@@ -144,25 +154,29 @@ export class DropdownComponent
...
 
   @HostBinding('class.align-top')
-  _alignTop: boolean;
+  _openUp: boolean;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:169]: Consider align class-name if renaming property...

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 24]
```diff
@@ -73,6 +74,10 @@ describe('DropdownComponent', () => {
       expect(buttonElement).toHaveText(spectator.component.placeholder, true);
     });
 
+    it('should have default open set', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:77]: Consider naming:
```suggestion
    it('should have default open direction set', () => {
```

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 25]
```diff
@@ -253,6 +258,36 @@ describe('DropdownComponent', () => {
...
+        setTimeout(() => {
+          const card = spectator.query('kirby-card');
+          const buttonRect = buttonElement.getBoundingClientRect();
+          const cardRect = card.getBoundingClientRect();
+
+          expect(cardRect.right).toEqual(buttonRect.right);
+          done();
+        }, openDelayInMs);
+      });
+    });
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:289]: Consider align tests if possible - any reason why we need to do this 2 different ways? 🤔

**Review by jakobe — CHANGES_REQUESTED** [review]
Almost there! 😄 💪 
Unfortunately `open()` and `close()` is part of the public and published API, so shouldn't change (or at least have a very good reason to...)
 I think we'll have to keep it as is and rename the direction input to `openDirection` or similar...
 Also please add the new property to the API map in dropdown-showcase component

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts (hunk 25 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts:289]: There was at the time :) -  I think it was related to how I initially set the direction in the intersectionObserver, so I needed the timeout. But now it works fine without

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 21 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: 👍  simplified

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 23 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 22 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: done

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 19 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Noted 👍  (and changed back again)

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 26]
```diff
@@ -257,12 +260,44 @@ export class DropdownComponent
     }
   }
 
+  private setHorizontalDirection(entry) {
+    // If card alignment is left, and the entry is cut off to the right by ${entry.boundingClientRect.right - entry.intersectionRect.right}px
+    // it is set to align to end in stead, and vice versa for right-aligned card
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:265]: Consider naming:
```suggestion
    // If popout direction is right, and the entry is cut off to the right by ${entry.boundingClientRect.right - entry.intersectionRect.right}px
    // it is set to pop out to the left instead, and vice versa for popout direction left
```

**Review by jakobe — CHANGES_REQUESTED** [review]

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 27]
```diff
@@ -247,6 +242,7 @@ export class DropdownComponent
         if (this.state !== OpenState.opening) {
           return;
         }
+        console.log('obserfver');
```
[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:245]: ouch.. not only a log statement, but a horribly spelled one

File: apps/cookbook/src/app/examples/dropdown-example/examples/right-aligned.ts [hunk 28]
```diff
@@ -6,7 +6,7 @@ const config = {
   [size]="size"
   placeholder="Right aligned (opens left)"
   [items]="['Item 1','Item 2','Item 3','Item 4','Item 5']"
-  open="left"
+  popOut="left"
```
[the remark this answers is not in the document]
Comment (flemcito) [reply apps/cookbook/src/app/examples/dropdown-example/examples/right-aligned.ts:9]: We'll settle on 1 word - looks nicer :)

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts [hunk 29]
```diff
@@ -74,8 +74,8 @@ describe('DropdownComponent', () => {
       expect(buttonElement).toHaveText(spectator.component.placeholder, true);
     });
 
-    it('should have default open set', () => {
-      expect(spectator.component.open).toEqual(HorizontalDirection.right);
+    it('should have default popout set', () => {
+      expect(spectator.component.popOut).toEqual(HorizontalDirection.right);
```
[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.spec.ts]: Fixed

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 30]
```diff
@@ -1,29 +1,44 @@
...
+  closed,
+  opening,
+  open,
+}
+
+export enum HorizontalDirection {
+  right = 'right',
+  left = 'left',
+}
+enum VerticalDirection {
```
[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Fixed

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 31]
```diff
@@ -40,8 +55,10 @@ import { CardComponent } from '../card/card.component';
 export class DropdownComponent
   implements AfterContentChecked, AfterViewInit, OnDestroy, ControlValueAccessor {
   static readonly OPEN_DELAY_IN_MS = 100;
-  private state: 'closed' | 'opening' | 'open' = 'closed';
+  private state = OpenState.closed;
   private hasConfiguredSlottedItems = false;
+  private _horizontal = HorizontalDirection.right;
+  private _vertical = VerticalDirection.down;
```
[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Agree with this. The `_horizontal` was the original name, so didnt think to change it. Fixed now

[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: outdated, went with 2nd suggestion :)

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 26 above)
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: fixed

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts [hunk 32]
```diff
@@ -1,29 +1,44 @@
...
 } from '@angular/core';
 import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
 
-import { ListItemTemplateDirective } from '../list/list.directive';
-import { ItemComponent } from '../item/item.component';
 import { CardComponent } from '../card/card.component';
+import { ItemComponent } from '../item/item.component';
+import { ListItemTemplateDirective } from '../list/list.directive';
+
+export enum OpenState {
```
[the remark this answers is not in the document]
Comment (flemcito) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts:28]: Yes, the test spec uses it also

File: libs/designsystem/src/lib/components/dropdown/dropdown.component.ts (hunk 26 above)
Comment (jakobe) [reply libs/designsystem/src/lib/components/dropdown/dropdown.component.ts]: Must have slipped a coomit/push...?

**Review by jakobe — APPROVED** [review]
🎉
